### ISSUE TITLE
Parsed DATE_ADD and DATE_SUB don't translate correctly when lacking column meta data

### ISSUE THREAD

**Issue by ghoshrahul** [jOOQ/jOOQ#12025]
### Expected behavior
On executing the SQL query, it should correctly add the interval specified in the query

### Actual behavior
Date add is not happening, it is returning the same result in a different format. The same is happening for days/month/year

### Steps to reproduce the problem
```java
String sqlQuery = " select users.id, users.created, ifnull(date_add(users.created, interval 2 hour), '0000-00-00 00:00:00') as TEST from users"
ResultQuery jooqQuery = DSL.using(dslContext.configuration()).parser().parseResultQuery(sqlQuery);
//Query generated: select users.id, users.created, ifnull((users.created + '+0 02:00:00.000000000'), '0000-00-00 00:00:00') as TEST from users
```
**Result**:

```
16:05:05.540 DEBUG org.jooq.tools.LoggerListener [http-nio-8080-exec-2] b27cc94c-f533-4664-95ac-0eaeff310345 - Fetched result           : +----+-------------------+--------------+
16:05:05.540 DEBUG org.jooq.tools.LoggerListener [http-nio-8080-exec-2] b27cc94c-f533-4664-95ac-0eaeff310345 -                          : |id  |created            |TEST          |
16:05:05.540 DEBUG org.jooq.tools.LoggerListener [http-nio-8080-exec-2] b27cc94c-f533-4664-95ac-0eaeff310345 -                          : +----+-------------------+--------------+
16:05:05.540 DEBUG org.jooq.tools.LoggerListener [http-nio-8080-exec-2] b27cc94c-f533-4664-95ac-0eaeff310345 -                          : |1   |2015-06-16T22:45:56|20150616224556|
```

### Versions

- jOOQ: 3.15.0-SNAPSHOT
- Java: 11
- Database (include vendor): MYSQL 5.7
- OS: Ubuntu 18.04
- JDBC Driver: com.mysql.jdbc.Driver

**Comment by lukaseder** [comment]
Thanks a lot for your report. I can reproduce this. Will look into this right away.

**Comment by ghoshrahul** [comment]
Hi Lukas,

The same issue can be found for MYSQL date_sub. Should I create another bug for this?

**Comment by lukaseder** [comment]
Thanks for the pointer. I think it's the same issue in this particular case. We don't distinguish internally between `DATE_ADD` and `DATE_SUB` yet, but solve this using positive or negative intervals. I'll update the title.

**Comment by lukaseder** [comment]
Ah I see. The problem is that the parser parses `DATE_ADD(a, b)` into `a + b`, which doesn't work (yet) if we don't have meta data. This would work correctly if you had written this instead:

```java
DSLContext ctx = DSL.using(MYSQL);
Meta meta = ctx.meta("create table users (id int, created timestamp)");
ctx.configuration().set((MetaProvider) () -> meta);
ctx.configuration().settings().setParseWithMetaLookups(ParseWithMetaLookups.THROW_ON_FAILURE);
String sqlQuery = " select users.id, users.created, date_add(users.created, interval 2 hour) as TEST from users";
ResultQuery jooqQuery = ctx.parser().parseResultQuery(sqlQuery);
System.out.println(jooqQuery);
```

The output is now:

```sql
select
  users.id,
  users.created,
  date_add(users.created, interval '+0 02:00:00.000000' day_microsecond) as TEST
from users
```

You'll probably get much better results in general, if jOOQ knows about your meta data. I'll think about how to fix this in the meantime. We now have a `DateAdd` expression tree element, which works better than the `a + b` version.

**Comment by ghoshrahul** [comment]
I have tried the method that you have mentioned in the previous comment. Yes it did work. I have a couple of questions though:-

1. Why is the ParseWithMetaLookups.THROW_ON_FAILURE required? When I tried with the default settings as well as using the meta data, it was not working. Why is it so?
2. Is there any way, where we can add the meta for all the tables in one go for the DSLContext? We are using the code generator by JOOQ. I understand `dslContext.meta()` accepts n number of table.

**Comment by lukaseder** [comment]
> 1\. Why is the ParseWithMetaLookups.THROW_ON_FAILURE required? When I tried with the default settings as well as using the meta data, it was not working. Why is it so?

Well, the default is to not use meta lookups at all (`OFF`), so you need to turn this on using `THROW_ON_FAILURE` or `IGNORE_ON_FAILURE`. The presence of meta data is not sufficient for this to work. Meta data are used for a variety of reasons in jOOQ, but using them in the parser creates a bit of overhead (especially when the meta data is read from JDBC), which is why the feature has been turned off when it was introduced.

> 2\. Is there any way, where we can add the meta for all the tables in one go for the DSLContext? We are using the code generator by JOOQ. I understand `dslContext.meta()` accepts n number of table.

Yes, you can pass the entire `Schema` reference to `DSLContext.meta(Schema)`. Check out the various overloads to see what's available. Also, you can call `Meta.filterXYZ()` to remove things again if you don't want the entire schema.

**Comment by lukaseder** [comment]
Looking into this now. While best results in the parser / translator are always achieved when providing meta data to the parser, additional improvements could be done by merging duplicate functionality between `DateAdd` and `Expression` implementations as documented here: https://github.com/jOOQ/jOOQ/issues/11563

That's quite the refactoring, which risks other types of regressions, so I will first explore a fix that ensures we don't use `Expression` in this particular case.

**Comment by lukaseder** [comment]
A potential fix for the specific parser problem is still a bit unsatisfactory.

The data type of the first argument can be `DATE` or `DATETIME` / `TIMESTAMP`, and without *knowing* this (via meta data), we can't produce the correct expression for all dialects. Making this configurable is overkill. The correct solution is to provide actual meta data.

A heuristic workaround would be to hard-code an assumption that the argument type is `TIMESTAMP`:

- This works correctly for half of the fields (if they actually *are* of type `TIMESTAMP`)
- Given the current translation implementations, this also works correctly for `DATE` fields for a lot of dialects, where the functionality doesn't differ between the types

But there could still be wrong translations for some input/output dialect pairs.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.16.0 and 3.15.2 (https://github.com/jOOQ/jOOQ/issues/12313)
